### ISSUE TITLE
Issue establishing ICE connectivity with Janus SFU as receiver only (Unity)

### ISSUE THREAD

**Issue by RyanWarwick** [microsoft/MixedReality-WebRTC#653]
**Describe the bug**
Hi, i am having issues setting up a receive-only PeerConnection to Janus SFU using the Unity packages. I have implemented my own signaling with Janus over Websockets and it is working fine, I am able to send video/audio to Janus SFU according to the logs and event messages. However, when attempting to create a new PeerConnection as a receiver-only, I can successfully accept and process Offer SDP messages, reply with an Answer and send ICE candidates to Janus but it claims there are no ICE remote-candidates gathered on the Janus side. Would greatly appreciate any leads!

**To Reproduce**
1. Create new instance of PeerConnection programmatically (not via Unity Editor)
2. Add Video and Audio receiver media lines programmatically 
3. Initialize PeerConnection programmatically
4. Establish Websocket connection with  Janus and process signalling
5. Process SDP Offer from Janus, reply with Answer
6. Process ICE candidates from Janus (?) but local PeerConnection ICE state does not change
7. No ICE remote-candidates found on Janus SFU side (viewable from Admin monitoring panel)

**Expected behavior**
Receive-only PeerConnection should be able to process Janus ICE Candidates (?) and Janus should be able to process PeerConnection ICE Candidates

**SDP messages**
_SDP messages are extremely helpful to diagnose a wide range of issues from connection not being established to audio/video issues and codecs issues._
<details open><summary>Offer message</summary>
<code>
 v=0
o=- 1611414440141835 1 IN IP4 1.2.3.4
s=VideoRoom 7135938545422058
t=0 0
a=group:BUNDLE 0 1
a=msid-semantic: WMS janus
m=audio 9 UDP/TLS/RTP/SAVPF 111
c=IN IP4 1.2.3.4
a=sendonly
a=mid:0
a=rtcp-mux
a=ice-ufrag:xteG
a=ice-pwd:5wlvJB7FoYRvT11iYfnRF7
a=ice-options:trickle
a=fingerprint:sha-256 ED:61:BA:95:CF:69:47:EC:94:2E:5D:30:F9:E9:39:6D:6D:EE:A1:00:A3:26:96:63:8A:6C:63:A4:54:2A:C8:25
a=setup:actpass
a=rtpmap:111 opus/48000/2
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=extmap:2 urn:ietf:params:rtp-hdrext:sdes:mid
a=rtcp-fb:111 transport-cc
a=msid:janus janusa0
a=ssrc:1658267727 cname:janus
a=ssrc:1658267727 msid:janus janusa0
a=ssrc:1658267727 mslabel:janus
a=ssrc:1658267727 label:janusa0
m=video 9 UDP/TLS/RTP/SAVPF 96 97
c=IN IP4 1.2.3.4
a=sendonly
a=mid:1
a=rtcp-mux
a=ice-ufrag:xteG
a=ice-pwd:5wlvJB7FoYRvT11iYfnRF7
a=ice-options:trickle
a=fingerprint:sha-256 ED:61:BA:95:CF:69:47:EC:94:2E:5D:30:F9:E9:39:6D:6D:EE:A1:00:A3:26:96:63:8A:6C:63:A4:54:2A:C8:25
a=setup:actpass
a=rtpmap:96 VP8/90000
a=rtcp-fb:96 ccm fir
a=rtcp-fb:96 nack
a=rtcp-fb:96 nack pli
a=rtcp-fb:96 goog-remb
a=extmap:2 urn:ietf:params:rtp-hdrext:sdes:mid
a=extmap:4 urn:3gpp:1-orientation
a=extmap:6 http://www.webrtc.org/experiments/rtp-hdrext/playout-delay
a=rtpmap:97 rtx/90000
a=fmtp:97 apt=96
a=ssrc-group:FID 3881950061 2032815598
a=msid:janus janusv0
a=ssrc:3881950061 cname:janus
a=ssrc:3881950061 msid:janus janusv0
a=ssrc:3881950061 mslabel:janus
a=ssrc:3881950061 label:janusv0
a=ssrc:2032815598 cname:janus
a=ssrc:2032815598 msid:janus janusv0
a=ssrc:2032815598 mslabel:janus
a=ssrc:2032815598 label:janusv0
</code>
</details>
<details open><summary>Answer message</summary>
<code>
v=0
o=- 4464691363840866046 2 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE audio video
a=msid-semantic: WMS janus
m=audio 9 UDP/TLS/RTP/SAVPF 111
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:f05R
a=ice-pwd:Q+hpl1gsQPmVWJIgLMe99gPJ
a=ice-options:trickle
a=fingerprint:sha-256 02:37:1D:FC:32:84:1A:F2:D1:99:24:7D:CE:49:CB:F4:00:39:E0:F8:C9:CE:9B:31:9D:31:99:35:9C:84:61:4B
a=setup:active
a=mid:audio
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=extmap:2 urn:ietf:params:rtp-hdrext:sdes:mid
a=recvonly
a=rtcp-mux
a=rtpmap:111 opus/48000/2
a=rtcp-fb:111 transport-cc
a=fmtp:111 minptime=10;useinbandfec=1
m=video 9 UDP/TLS/RTP/SAVPF 96 97
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:f05R
a=ice-pwd:Q+hpl1gsQPmVWJIgLMe99gPJ
a=ice-options:trickle
a=fingerprint:sha-256 02:37:1D:FC:32:84:1A:F2:D1:99:24:7D:CE:49:CB:F4:00:39:E0:F8:C9:CE:9B:31:9D:31:99:35:9C:84:61:4B
a=setup:active
a=mid:video
a=extmap:6 http://www.webrtc.org/experiments/rtp-hdrext/playout-delay
a=extmap:2 urn:ietf:params:rtp-hdrext:sdes:mid
a=recvonly
a=rtcp-mux
a=rtpmap:96 VP8/90000
a=rtcp-fb:96 goog-remb
a=rtcp-fb:96 ccm fir
a=rtcp-fb:96 nack
a=rtcp-fb:96 nack pli
a=rtpmap:97 rtx/90000
a=fmtp:97 apt=96
</code>
</details>

**Screenshots**
_If applicable, add screenshots to help explain your problem._

**Environment**
_Please fill the information for **each peer** if different_
 - MR-WebRTC version unity package v.2.0.2
 - Platform: Unity Editor
 - Architecture: Unity Editor is x64 on Windows
 - Unity version : 2020.2.1f1
 - Target device: Windows Desktop (for now)

**Additional context**
Sample ICE Candidate from Janus : candidate:2 1 udp 1679819007 1.2.3.4 48658 typ srflx raddr 172.31.47.70 rport 48658
Sample ICE Candidate from local PeerConnection :
candidate:3460887983 1 udp 2122260223 192.168.1.64 50205 typ host generation 0 ufrag f05R network-id 1 network-cost 10
candidate:945327227 1 udp 1686052607 5.6.7.8 56797 typ srflx raddr 192.168.1.64 rport 50205 generation 0 ufrag f05R network-id 1 network-cost 10
candidate:2160789855 1 tcp 1518280447 192.168.1.64 65141 typ host tcptype passive generation 0 ufrag f05R network-id 1 network-cost 10

**Comment by djee-ms** [comment]
It sounds like the issue is ICE candidates either a) not being generated by the Unity lib (never heard of that), or b) not being transported correctly by your custom signalling, or c) not being handled correctly on reception. It would be nice to be able to narrow down where the issue is. Do you see the ICE candidates generated on the sender side (Unity)? Do you see them arriving on the Janus side? In the description you say the issue is Janus -> Unity, but in the repro step 6. you also seem to say that Unity -> Janus may be an issue. If both sides don't work, that sounds like a signalling and/or networking issue.

**Comment by RyanWarwick** [comment]
Thanks for the reply. To clarify, what i meant is that in terms of media, it would be Janus -> Unity. However, in order for this to happen, Janus will send an SDP Offer + ICE Candidates to Unity over the custom signalling. I am able to receive this and generate an SDP Answer + my own ICE Candidates, which i send back to Janus. Janus is able to accept the SDP Answer. However, it does not seem to accept the ICE Candidates.

At the same time, when I run _nativePeer.AddCandidate on the ICE Candidates received from Janus, the ICE connectivity state for the PeerConnection does not update to a checking state. 

Since my implementation works for sending media from Unity -> Janus works and i have compared my signalling JSON content to that of the Janus.js library (which are identical), it is most likely not a signalling issue. Based on logs on the Janus server instance, it is definitely receiving all the SDP/ICE messages but not processing them, so i dont think networking issues are at play either. This seems to be the case for Unity side too, where it can receive the ICE Candidates but does not process them (?) so I am confused as to which side could be the culprit. 

Attached are sample SDP / ICE messages for the process described above. Thanks again.

**Comment by RyanWarwick** [comment]
UPDATE
---------
Upon further tweaking, it seems that Janus is picky on the exact fields being sent over. I have resolved the signaling portion such that Janus receives the ICE Candidates successfully now. However, it fails when processing them and attempting the connection with the Unity peer. Since this is most likely a connectivity issue, could it be attributed to the lack of response/change in ICE state when calling _nativePeer.AddIceCandiate(<janus ice candidate>)?

**Comment by djee-ms** [comment]
I don't know Janus, if it fails to process ICE messages then there's a good chance indeed that the connection will fail, since the peers won't know how to reach each other to establish a direct P2P connection.

**Comment by RyanWarwick** [comment]
Hmm, is _nativePeer.AddIceCandidate() supposed to result in a change in ice connectivity state everytime? As indicated in the debug logs? Like 'ICE Connectivity : Checking'?
I am trying to determine if it is an issue with Janus itself or not, as calling _nativePeer.AddIceCandidate() as a sender peer has such an output. But calling it as a receiver does not have any output

**Comment by djee-ms** [comment]
`_nativePeer.AddIceCandidate()` is passing 1 ICE candidate received from the remote peer to the internal Google implementation of the local peer. My understanding of Google's code is that the implementation will process ICE candidates as they arrive, and _in parallel_ will try to establish a P2P connection. Even after the P2P connection was established, a new ICE candidate _may_ provide a better route (as decided by that implementation) and therefore switch to a new P2P route. I am not sure about the exact event(s) that change the ICE state reported, I think "checking" corresponds to the first candidate sent or processed, and "completed" to the last one, but I didn't look in detail.

**Comment by RyanWarwick** [comment]
I've finally been able to establish connectivity as a Subscriber to Janus. I went to do further reading into AddIceCandidate() under Google's documentation as you pointed out. The issue was that mixed-reality-webrtc's SDP messages refer to the sdpMid as 0/1, while Janus refers to it as Audio/Video. So in order for them to establish the connection, the sdpMid sent from one side's SDP message must be identical to the sdpMid from its ICE Candidate message as well. I had manipulated the incoming SDP Offer data from Janus (from Audio/Video -> 0/1) for mixed-reality-webrtc to process it, but had not done so to match the SDP Answer + ICE Candidate on the outgoing. Hope this helps others. Thanks again!

**Comment by djee-ms** [comment]
Yes I think if I remember correctly some time ago we had such assumption on `mid` because we don't have any other reliable way to auto-match the transceivers (in Unity). Although I thought this assumption had been lifted and the bug fixed (see #295).

**Comment by RyanWarwick** [comment]
Hi @djee-ms , apologies for the late reply. I had been away from this project during the recent Lunar New Year. Upon further review of my code, it seems that there was some confusion on my end regarding the setting of mid value ( 0/1 vs audio/video ). I had looked up some other related issues and fixes that were pertaining to the pre-bug fix and probably mixed up different values when configuring SDP Answer vs ICE Candidates. You are right, mixedreality-webrtc Unity does accept both sets of values. 

However, something to take note from what I have tested is that Janus sends the mid as audio/video. Unity side is able to accept these values and establish ice connectivity/receive and display remote video. Everything works fine. If I were to set it to 0/1 on the Unity side instead, I noticed that it is still able to establish ice connectivity and receive remote video, but is unable to display it. It shows the set of "infinity" values on the remote peer window, which may be a concern.
